src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts

```diff
@@ -122,58 +125,99 @@ export class InstructorCourseEnrollPageComponent implements OnInit {
-                const existingStudentTable: Handsontable = this.hotRegisterer.getInstance(this.existingStudentsHOT);
-                this.loadExistingStudentsData(existingStudentTable, this.existingStudents);
-              }
-              this.isExistingStudentsPresent = true;
-            });
-          },
-          error: (resp: ErrorMessageOutput) => {
-            this.enrollErrorMessage = resp.error.message;
-          },
+    );
+
+    const handledRequests: Observable<Students> = this.handleTimeoutForEnrollRequest(enrollRequests);
+    handledRequests.subscribe({
+      next: (resp: Students) => {
+        enrolledStudents.push(...resp.students);
+        this.remainingStudentChunks = this.remainingStudentChunks.slice(1, this.remainingStudentChunks.length);
+      },
+      complete: () => {
+        this.showEnrollResults = true;
+        this.statusMessage.pop(); // removes any existing error status message
+        this.statusMessageService.showSuccessToast('Enrollment successful. Summary given below.');
+
+        const allStudentEntries: StudentEnrollRequest[] = [];
+        this.allStudentChunks.forEach((chunk: StudentEnrollRequest[]) => allStudentEntries.push(...chunk));
+
+        this.enrollResultPanelList =
+            this.populateEnrollResultPanelList(this.existingStudents, enrolledStudents, allStudentEntries);
+
+        this.studentService.getStudentsFromCourse({ courseId: this.courseId }).subscribe((resp: Students) => {
+          this.existingStudents = resp.students;
+          if (!this.isExistingStudentsPanelCollapsed) {
+            const existingStudentTable: Handsontable = this.hotRegisterer.getInstance(this.existingStudentsHOT);
+            this.loadExistingStudentsData(existingStudentTable, this.existingStudents);
+          }
+          this.isExistingStudentsPresent = true;
         });
+      },
+      error: (resp: ErrorMessageOutput) => {
+        this.enrollErrorMessage = resp.error.message;
+      },
+    });
   }
 
+  private handleTimeoutForEnrollRequest(prevRequests: Observable<Students>): Observable<Students> {
+
+    return prevRequests.pipe(finalize(() => this.isEnrolling = false))
+        .pipe(takeWhile(() => this.remainingStudentChunks.length > 0))
+        .pipe(catchError(() => {
+          // First request in the list caused the timeout
+          const currentLongRequest: StudentEnrollRequest[] = this.remainingStudentChunks[0];
+
+          // Break the long request into half
+          const studentEnrollRequest1: StudentEnrollRequest[] =
+              currentLongRequest.slice(0, currentLongRequest.length / 2);
+
+          const studentEnrollRequest2: StudentEnrollRequest[] =
+              currentLongRequest.slice(currentLongRequest.length / 2, currentLongRequest.length);
+
+          this.remainingStudentChunks =
+              this.remainingStudentChunks.filter((chunk: StudentEnrollRequest[]) =>
+                  chunk !== currentLongRequest);
```

Comment:
Another thing about this entire filtering logic: would it even not only filter the first element like line 190 suggests?
If it ever filters more than two elements from the array, how should we regard it?

Reply:
this line is trying to remove the long request and replace it with 2 smaller ones. I will change it to ```shift()``` to make sure only 1 chunk being removed

Reply:
Awesome.
Some tips I have regarding this: try to explore the existing functions of a type (in any languages), especially the standard library ones; they are added in because they either improve efficiency or readability and are practical (unless the function is deemed notoriously unnecessary in the community). Try to use them as much as possible instead of hand crafting what they already do.